sdk/core/core-rest-pipeline/src/retryStrategies/failoverRetryStrategy.ts

```diff
@@ -0,0 +1,153 @@
+   */
+  retryCount: number;
+  /**
+   * The policy should wait until `retryAt` has elapsed to try this host.
+   */
+  retryAt?: Date;
+}
+declare interface FailoverHosts {
+  index: number;
+  hostStates: FailoverHostState[];
+}
+
+/**
+ * Default behavior for generating hosts for failover.
+ *
+ * If the request method is GET, HEAD, or OPTIONS, it cycles through `readHosts`. Otherwise, it cycles through `writeHosts`.
+ */
+export function defaultFailoverHostGenerator(
+  readHosts: string[] = [],
+  writeHosts: string[] = []
+): FailoverHostGenerator {
+  const readHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: readHosts.map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+  const writeHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: writeHosts.map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+  return function* (retryState, options) {
+    while (retryState) {
+      const response = retryState.response;
+      const retryDelayInMs = options.retryDelayInMs ?? DEFAULT_CLIENT_RETRY_INTERVAL;
+      const maxRetryDelayInMs = options.maxRetryDelayInMs ?? DEFAULT_CLIENT_MAX_RETRY_INTERVAL;
+      if (!response) {
+        return;
+      }
+
+      // Select correct list of hosts for this request
+      const isSafeMethod = ["GET", "HEAD", "OPTIONS"].includes(response.request.method);
+      const currentStateList = isSafeMethod ? readHostStateList : writeHostStateList;
+
+      if (currentStateList.hostStates.length === 0) {
+        yield undefined;
+        continue;
+      }
+      const hostState = currentStateList.hostStates[currentStateList.index];
+
+      const requestFireTime = Math.max(hostState.retryAt?.valueOf() ?? -Infinity, Date.now());
+
+      retryState = yield hostState;
+
+      const exponentialDelay = Math.min(
+        retryDelayInMs * 2 ** hostState.retryCount,
+        maxRetryDelayInMs
+      );
+      const nextRetryAt = new Date(requestFireTime + exponentialDelay);
+      hostState.retryAt = nextRetryAt;
+
+      currentStateList.index = (currentStateList.index + 1) % currentStateList.hostStates.length;
+      hostState.retryCount++;
+    }
+  };
+}
+
+/**
+ *
+ * @param options - Lists of fallback endpoints for read/write operations
+ * @returns An exponential retry strategy with support for multiple endpoints
+ */
+export function failoverRetryStrategy(options: DefaultRetryPolicyOptions): RetryStrategy {
+  const generatorMap = new WeakMap<PipelineRequest, ReturnType<FailoverHostGenerator>>();
+  return {
+    name: "failoverRetryStrategy",
+    retry(state) {
+      const response = state.response;
+      // Only fallback to another host on a 5xx response
+      if (!(response && 500 <= response.status && response.status < 600)) {
+        return { skipStrategy: true };
+      }
+
+      if (!options.failoverHostGenerator) {
```

Comment:
Perhaps also worth having the default retry policy check this and not include the strategy when it is unset

Reply:
Done!